### PR TITLE
feat: fetch latest gradle release if no version is given

### LINKED ISSUE #2650: Should AddGradleWrapper default to latest version if not provided?
Currently, the `AddGradleWrapper` recipe expects `version` as a mandatory parameter instead of defaulting to the latest version

Gradle supports this natively as running `gradle wrapper` defaults to the latest version if not provided

Additional context: 

I am not sure if this fits this project. 

I understand that it could become a maintenance overhead if the latest version is maintained manually. We could query https://services.gradle.org/distributions/ to find the latest instead of keeping it hard-coded but that means that the recipe would not be be self-contained/work offline.

### PR DESCRIPTION
**Pull request by yeikel**
Closes #2650

### CODE REVIEW

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java [hunk 1]
```diff
@@ -108,6 +101,67 @@ void addWrapper() {
...
+                  output.set(wrapperProperties.printAll());
+
+              }),
+
+          buildGradle(""),
+
+          dir(
+            "gradle/wrapper",
+            properties(null, output, spec -> spec.path(Paths.get("gradle-wrapper.properties")))
+          ));
```
Comment (yeikel) [inline rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:150]: This assertion could be better if we had a mechanism to mock the response. This value is dynamic so I did not find any other better way to verify it other than this

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java [hunk 2]
```diff
@@ -57,8 +59,10 @@ public String getDescription() {
     }
 
     @Option(displayName = "New version",
-            description = "An exact version number or node-style semver selector used to select the version number.",
-            example = "7.x")
+            description = "An exact version number or node-style semver selector used to select the version number. It defaults to the latest RELEASE",
```
Comment (yeikel) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java:62]: This description might be confusing. I am open for suggestions

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java [hunk 3]
```diff
@@ -88,6 +92,12 @@ public AddGradleWrapper(String version, String distribution) {
         this.distribution = distribution;
     }
 
+    @LoathingOfOthers("JavaDoc")
+    @JsonCreator
+    public AddGradleWrapper() {
```
Comment (shanman190) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java:97]: So the typical way of handling this would be to check if version is null or empty in the `getVisitor()` method. If null or empty, set version to RELEASE and pass into the same existing flow.

The `VersionComparator` class already has all of the necessary logic to select the latest release version, so we don't have to make a separate alternative HTTP invocation.

By doing this that means, that we don't need a separate constructor either.

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java [hunk 4]
```diff
@@ -108,6 +101,67 @@ void addWrapper() {
...
+    @Test
+    void addWrapperWithReleaseTag() {
+
+        AtomicReference<String> output = new AtomicReference<>("");
+        rewriteRun(
+          spec ->
+            spec
+              .recipe(
+                //language=yaml
+                new ByteArrayInputStream(
```
Comment (shanman190) [inline rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:114]: With my other comment, this simply becomes creating the recipe with all null constructor arguments.

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java (hunk 1 above)
Comment (shanman190) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:150]: So you can do the assertion with the `Assertions.properties` call. And I believe that there is a way to mock the dynamic versions now.

@jkschneider, do you have any insight on that mocking support?

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java (hunk 3 above)
Comment (yeikel) [reply rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java:97]: When I tried it, this block

```java

GradleVersion gradleVersion = allVersions.stream()
                            .filter(v -> versionComparator.isValid(null, v.version))
                            .max((v1, v2) -> versionComparator.compare(null, v1.version, v2.version))
                            .orElseThrow(() -> new IllegalStateException("Expected to find at least one Gradle wrapper version to select from."));

```

Failed to parse it as it did not find any matching release from `allVersions` 

``allVersions`` is an array of versions but I did not find any attribute stating that it would be the latest or not

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java (hunk 4 above)
Comment (yeikel) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:114]: Wouldn't this be confusing? 

And, would we expect people to pass them while building them? 

With the current setup, it seems that the recipe cannot be executed from the command line as it needs this formal configuration.

Comment (shanman190) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:114]: So the short answer, is yes. The build plugins both understand required vs optional parameters, so if all parameters are optional then they are able to just provide null or empty values. This enables construction of recipes that have many parameters that are all optional or of recipes which have some required and some optional without having to specify each of those parameters.

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java (hunk 1 above)
Comment (jkschneider) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:150]: Yes, I'll put up a polish commit showing how to do it.

**Comment by shanman190** [comment]
@yeikel, apologies about this, but just realized that `RELEASE` is incorrect and it should be `latest.release`. I can say that that with that value it does work. 

Ultimately I ended up with something like this as an reference:
```
diff --git a/rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java b/rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java
index aeac69f77..8dbec6269 100644
--- a/rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java
+++ b/rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java
@@ -58,7 +58,9 @@ public String getDescription() {

     @Option(displayName = "New version",
             description = "An exact version number or node-style semver selector used to select the version number.",
-            example = "7.x")
+            example = "7.x",
+            required = false)
+    @Nullable
     String version;

     @Option(displayName = "Distribution type",
@@ -76,7 +78,7 @@ public String getDescription() {

     @Override
     public Validated validate(ExecutionContext ctx) {
-        return super.validate(ctx).and(GradleWrapper.validate(ctx, version, distribution, gradleWrapper));
+        return super.validate(ctx).and(GradleWrapper.validate(ctx, version == null ? "latest.release" : version, distribution, gradleWrapper));
     }

     //NOTE: Using an explicit constructor here due to a bug that surfaces when running JavaDoc.
```

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java (hunk 1 above)
Comment (yeikel) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:158]: @shanman190 I believe that I applied all your feedback in my latest commits :) 

@jkschneider Did you get a chance to set up the example you mentioned?

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java [hunk 5]
```diff
@@ -77,7 +79,7 @@ public String getDescription() {
 
     @Override
     public Validated validate(ExecutionContext ctx) {
-        return super.validate(ctx).and(GradleWrapper.validate(ctx, version, distribution, gradleWrapper));
+        return super.validate(ctx).and(GradleWrapper.validate(ctx, version == null ? "latest.release" : version, distribution, gradleWrapper));
```
Comment (shanman190) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java:82]: Thinking about this again, it would probably be better to use `org.openrewrite.internal.StringUtils.isBlank(version)` here to ensure that it properly handles null or empty strings.

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java (hunk 4 above)
Comment (shanman190) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:119]: I think it might be easier to understand this if we were to use the java constructor to create these instances. This also helps to act as a verifiable test for my other comment as it seems the YAML recipe version is coercing the blanks to null which doesn't happen everywhere the same way.

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java (hunk 1 above)
Comment (shanman190) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java:158]: So the `org.openrewrite.properties.Assertions.properties(..)` method handles the directory pathing so that you don't have to. Also not sure of the value of capturing the output instead of just hard coding the assertion and seeing what Jon has in store from a mocking standpoint when he submits his polish commit on top.

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java (hunk 5 above)
Comment (yeikel) [reply rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java]: Thank you. I addressed that with 23a9ce7641f3efa35e9c68daab6c6001eafa18ac

File: rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java (hunk 1 above)
Comment (yeikel) [reply rewrite-gradle/src/test/java/org/openrewrite/gradle/AddGradleWrapperTest.java]: I hard-coded it to `7.6` for now and the PR is "ready". This test is bound to fail in the future when a new version comes out so we can wait for Jon to share an example for the mocking before we merge :)

**Review by shanman190 — APPROVED** [review]
@jkschneider, do you want to merge this since you were saying that you were going to add a polishing commit?

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java [hunk 6]
```diff
@@ -58,8 +58,10 @@ public String getDescription() {
     }
 
     @Option(displayName = "New version",
-            description = "An exact version number or node-style semver selector used to select the version number.",
-            example = "7.x")
+            description = "An exact version number or node-style semver selector used to select the version number. It defaults to the latest RELEASE",
```
Comment (sambsnyd) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java:61]:
```suggestion
            description = "An exact version number or [dependency version selector](https://docs.openrewrite.org/reference/dependency-version-selectors). Defaults to the latest release version.",
```

**Comment by sambsnyd** [comment]
I'm not sure what, specifically, Jonathan had in mind. But I started looking at this and added the beginnings of mocking of gradle wrapper artifacts. Will finish it later. Sorry for messing up the commit history with a sloppy rebase, but it should squash-merge just fine all the same.
